Improvements to JITLink, to ease transition

vtjnash opened the pull request:
Seeing what this will look like, since it has a number of features (delayed compilation, concurrent compilation) that are starting to become important, so it would be nice to switch to only supporting one common implmenetation of memory management.

Refs #50248

I am expecting https://github.com/llvm/llvm-project/issues/63236 may cause problems, since we reconfigured some CI machines to hide that issue, but it is still likely relevant.

giordano commented on src/codegen.cpp:
```diff
@@ -10243,9 +10243,6 @@ extern "C" void jl_init_llvm(void)
 #if defined(JL_USE_INTEL_JITEVENTS) || \
     defined(JL_USE_OPROFILE_JITEVENTS) || \
     defined(JL_USE_PERF_JITEVENTS)
-#ifdef JL_USE_JITLINK
-#pragma message("JIT profiling support (JL_USE_*_JITEVENTS) not yet available on platforms that use JITLink")
-#else
```
My understanding is that this is still true in llvm 18 (and at the moment we support up to 17)?  https://github.com/llvm/llvm-project/pull/83957 isn't in any stable version yet (presumably will be in LLVM 19).

gbaraldi replied:
It's likely this backports quite cleanly to 18 though

vtjnash replied:
It is no longer relevant though, since they are (increasingly) available with each new version of LLVM

vchuravy commented:
PPC got further than I expected.

gbaraldi commented:
Both are on LLVM 18 so I wonder if we should skip a release with 17, since the big changes that cause issues for people are already there (no opaque ptrs)

giordano commented:
>  I wonder if we should skip a release with 17

I don't think we're committed to have each major LLVM version in a Julia minor version: Julia v1.5 had LLVM 9, v1.6 had LLVM 11.

vtjnash committed:
Improvements to JITLink

Seeing what this will look like, since it has a number of features
(delayed compilation, concurrent compilation) that are starting to
become important, so it would be nice to switch to only supporting one
common implementation of memory management.

Refs #50248

I am expecting https://github.com/llvm/llvm-project/issues/63236 may
cause some problems, since we reconfigured some CI machines to minimize
that issue, but it is still likely relevant.

vtjnash committed:
rewrite catchjmp asm to use normal relocations instead of manual editing

vchuravy commented on src/jitlayers.cpp:
```diff
@@ -1973,42 +1941,75 @@ StringRef JuliaOJIT::getFunctionAtAddress(uint64_t Addr, jl_callptr_t invoke, jl
 }
 
-void JuliaOJIT::RegisterJITEventListener(JITEventListener *L)
+void JuliaOJIT::enableOProfileJITEventListener()
```
Do we want to just remove OProfile?

vtjnash replied:
oops, no I just forgot to move the implementation here

gbaraldi commented on src/jitlayers.cpp:
```diff
@@ -281,18 +288,40 @@ static jl_callptr_t _jl_compile_codeinst(
+        if (!decls.specFunctionObject.empty())
+            NewDefs.push_back(decls.specFunctionObject);
+    }
+    auto Addrs = jl_ExecutionEngine->findSymbols(NewDefs);
```
This is the thing we discussed that makes the memory a little better right?

vtjnash replied:
Yes, this helps because it ensures it does the whole graph at once (behind the codegen lock). We can also later turn off the logic that tries to filter the graph, but this already should give the performance effect
